neutrinos/kipp.H

```diff
@@ -0,0 +1,177 @@
+// This routine calculates the neutrino cooling rates using the approximations from Kippenhahn et al. 1989
+// using autodiff
+
+#ifndef KIP_NEUT_H
+#define KIP_NEUT_H
+
+#define _USE_MATH_DEFINES
+
+#include<iostream>
+#include<AMReX_REAL.H>
+#include<fundamental_constants.H>
+#include<microphysics_autodiff.H>
+
+using namespace amrex::literals;
+
+template<typename number_t>
+AMREX_GPU_HOST_DEVICE AMREX_INLINE
+// pair annihilation
+number_t kip_pair(const number_t& temp, const amrex::Real& rho,
+                  const number_t& abar, const number_t& zbar){
+    number_t T9 = temp * 1.0e-9_rt;
+    // eq. 18.81 - fixed using woosley lecture notes
+    if (T9 < 2.0e0_rt){
+        return (4.9e18_rt/rho) * amrex::Math::powi<3>(T9) * admath::exp(-11.86/T9);
+    }
+    else{
+        return (3.2e15_rt/rho) * amrex::Math::powi<9>(T9);
+    }
+}
+
+template<typename number_t>
+AMREX_GPU_HOST_DEVICE AMREX_INLINE
+// photoneutrinos
+number_t kip_phot(const number_t& temp, const amrex::Real& rho,
+                  const number_t& abar, const number_t& zbar){
+    number_t T9 = temp * 1.0e-9_rt;
+    // eq. 18.82
+    number_t ep_1 = (1.103e13_rt/rho) * amrex::Math::powi<9>(T9) * admath::exp(-5.93e0_rt / T9);
+    number_t ep_2 = 0.976e8_rt * amrex::Math::powi<8>(T9) / (1.0e0_rt + 4.2e0_rt * T9);
+    number_t rho_bar = 6.446e-6_rt * rho / (T9 + 4.2e0_rt * amrex::Math::powi<2>(T9));
+    number_t mu_e = abar / zbar;
+
+    return ep_1 + ep_2 / (mu_e + rho_bar);
+}
+
+template<typename number_t>
+AMREX_GPU_HOST_DEVICE AMREX_INLINE
+// plasmaneutrinos
+number_t kip_plas(const number_t& temp, const amrex::Real& rho,
+                  const number_t& abar, const number_t& zbar){
+    number_t n_e = rho / (C::m_u * abar/zbar);
+    number_t T9 = temp * 1.0e-9_rt;
+    // assuming non-degeneracy
+    // 18.83 replaced by aprox in woosley lecture notes
+    number_t w0 = admath::sqrt((4.0e0_rt * M_PI * amrex::Math::powi<2>(C::q_e) * n_e) / C::m_e);
+    number_t gamma = (C::hbar * w0) / (C::k_B * temp);
+    number_t lambda = (C::k_B * temp) / (C::m_e * amrex::Math::powi<2>(C::c_light));
+    // 18.85
+   if (gamma <= 1.0e0_rt) {
+       // return (7.4e21_rt * amrex::Math::powi<6>(C::hbar * w0/(C::m_e * amrex::Math::powi<2>(C::c_light)))
+       //         * amrex::Math::powi<3>(lambda))/rho;
+        auto epsilon_p = (C::hbar * w0) / (C::m_e * C::c_light * C::c_light);
+        return (7.4e21_rt * admath::pow(epsilon_p, 6.0_rt) * admath::pow(lambda, 3.0_rt)) / rho;
```

Comment:
When assigning an intermediate expression to a value, either the destination type should be `number_t` or the expression should be wrapped with `autodiff::eval()`, to force the expression template to be evaluated.
```suggestion
        number_t epsilon_p = (C::hbar * w0) / (C::m_e * C::c_light * C::c_light);
        return (7.4e21_rt * amrex::Math::powi<6>(epsilon_p) * amrex::Math::powi<3>(lambda)) / rho;
```